contentcuration/contentcuration/serializers.py

```diff
@@ -557,16 +564,41 @@ def retrieve_metadata(self, node):
                 "has_changed_descendant": node.changed,
             }
 
+    def retrieve_source_channel(self, node):
+        # TODO(davidhu): This incurs an extra DB query -- how can we avoid?
+        # Some dirty / hacky command line profiling,
+        # `time for i in {1..100}; do curl http://.../api/get_nodes_by_ids/...; done`,
+        # reveals that this incurs about a 5% slowdown for that call (with
+        # about 32 nodes in the list).
+        #
+        # I also tried to get
+        # github.com/dmclain/django-debug-toolbar-line-profiler to work, but
+        # its panel page would show a 500 when I loaded up
+        # /api/get_nodes_by_ids/...
+        source_node = ContentNode.objects.filter(
```

Comment:
Each node should also have an `original_channel_id` on the [model](https://github.com/learningequality/studio/blob/develop/contentcuration/contentcuration/models.py#L589), which might help if the original node has been removed from the channel (might also be why some nodes are disassociated from channels). Also, could you add a filter to make sure the channel hasn't been deleted? We probably don't want any deleted channels to show up as segments on the clipboard

Reply:
Ah ha, using `Channel.objects.get(pk=node.original_channel_id)` seems to fetch us the correct first-origin channel for each node, and gets rid of the "Unknown channel" completely.

So I'm wondering what the difference between that is and the existing `ContentNodeSerializer.retrieve_original_channel()`. Here's the implementations respectively:

```
    def retrieve_source_channel(self, node):
        # NOTE: This implementation has been changed to retrieve the first-origin channel, not the immediate channel copied from
        channel_id = node.original_channel_id
        channel = channel_id and Channel.objects.get(pk=channel_id)
        return {"id": channel.pk, ...}

    def retrieve_original_channel(self, node):
        original = node.get_original_node()
        channel = original.get_channel() if original else None
        return {"id": channel.pk, ...}
```